### PR TITLE
[FLINK-20002] Add a StreamExecutionEnvironment#getExecutionEnvironment(Configuration) method

### PR DESCRIPTION
**Pull request by dawidwys**
## What is the purpose of the change

This PR adds a method to instantiate a StreamExecutionEnvironment with a given `Configuration`. It is a shortcut for 
```
StreamExecutionEnvironment env = new *StreamExecutionEnvironment(Configuration);
env.configure(Configuration);
```

## Verifying this change

Added tests in:
* `StreamExecutionEnvironmentComplexConfigurationTest`
* Changed the way environment is instantiated in `DataStreamBatchExecutionITCase`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/LocalStreamEnvironment.java [hunk 1]
```diff
@@ -66,6 +66,9 @@ private static Configuration validateAndGetConfiguration(final Configuration con
 		final Configuration effectiveConfiguration = new Configuration(checkNotNull(configuration));
 		effectiveConfiguration.set(DeploymentOptions.TARGET, "local");
 		effectiveConfiguration.set(DeploymentOptions.ATTACHED, true);
+		if (!configuration.getOptional(CoreOptions.DEFAULT_PARALLELISM).isPresent()) {
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/LocalStreamEnvironment.java:69]: Why is this needed?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java [hunk 2]
```diff
@@ -2043,9 +2044,23 @@ public void addOperator(Transformation<?> transformation) {
...
 	 */
 	public static StreamExecutionEnvironment getExecutionEnvironment() {
+		return getExecutionEnvironment(new Configuration());
+	}
+
+	/**
+	 * Creates an execution environment that represents the context in which the
+	 * program is currently executed. If the program is invoked standalone, this
+	 * method returns a local execution environment, as returned by
+	 * {@link #createLocalEnvironment(Configuration)}.
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java:2054]: We should mention that the `Configuration` is (potentially) "layered" on top of a base configuration or merged with it.

File: flink-tests/src/test/java/org/apache/flink/api/datastream/DataStreamBatchExecutionITCase.java [hunk 3]
```diff
@@ -127,17 +130,17 @@ public void batchSumSingleResultPerKey() throws Exception {
 	}
 
 	private StreamExecutionEnvironment getExecutionEnvironment() {
-		final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
 
 		Configuration config = new Configuration();
 		config.set(ExecutionOptions.RUNTIME_MODE, RuntimeExecutionMode.BATCH);
-		env.configure(config, DataStreamBatchExecutionITCase.class.getClassLoader());
+		config.set(CoreOptions.DEFAULT_PARALLELISM, 1);
```
Comment (aljoscha) [inline flink-tests/src/test/java/org/apache/flink/api/datastream/DataStreamBatchExecutionITCase.java:136]: So you prefer `config.set(CoreOptions.DEFAULT_PARALLELISM, 1);` to `env.setParallelism(1)`? 😅

**Review by aljoscha — CHANGES_REQUESTED** [review]
The changes look good! But I had some suggestions for Javadocs.

File: flink-tests/src/test/java/org/apache/flink/api/datastream/DataStreamBatchExecutionITCase.java (hunk 3 above)
Comment (dawidwys) [reply flink-tests/src/test/java/org/apache/flink/api/datastream/DataStreamBatchExecutionITCase.java:136]: Not necessarily. It is more to verify that we can pass parallelism that way as well. I can revert.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/LocalStreamEnvironment.java (hunk 1 above)
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/LocalStreamEnvironment.java:69]: I did it to keep the behaviour of the `LocalStreamEnvironment` class. It says:

```
<p>When this environment is instantiated, it uses a default parallelism of {@code 1}. The default parallelism can be set via {@link #setParallelism(int)}.
```
and it was explicitly calling `setParallelism(1)` irrespective of the `DEFAULT_PARALLELISM.defaultValue()`.

If you think its fine to just use the `DEFAULT_PARALLELISM.defaultValue()`, which is `1` btw, I can remove all the mingling with parallelism in here.

Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/LocalStreamEnvironment.java]: It should be fine to simplify it. 👌

**Review by aljoscha — APPROVED** [review]
👍
